syntax = "proto3";

package audiobook;

option swift_prefix = "AB";

// ============================================================================
// CONVERSION SERVICE
// ============================================================================

service ConversionService {
  // Start a conversion with real-time progress streaming
  rpc ConvertBook(ConversionRequest) returns (stream ConversionProgress);
  
  // Cancel an ongoing conversion
  rpc CancelConversion(CancelRequest) returns (CancelResponse);
  
  // Get current conversion status
  rpc GetConversionStatus(Empty) returns (ConversionStatus);
  
  // Preview voice with sample text
  rpc PreviewVoice(PreviewRequest) returns (stream AudioChunk);
}

service ModelService {
  // Get status of loaded models
  rpc GetModelStatus(Empty) returns (ModelStatus);
  
  // Load a model with progress streaming
  rpc LoadModel(LoadModelRequest) returns (stream LoadProgress);
  
  // Unload models to free memory
  rpc UnloadModel(UnloadModelRequest) returns (Empty);
  
  // Stream model status changes
  rpc StreamModelStatus(Empty) returns (stream ModelStatus);
}

service LibraryService {
  // List all audiobooks in library
  rpc ListBooks(ListBooksRequest) returns (ListBooksResponse);
  
  // Get details for a specific book
  rpc GetBook(GetBookRequest) returns (Book);
  
  // Delete a book from library
  rpc DeleteBook(DeleteBookRequest) returns (Empty);
  
  // Stream audio for playback
  rpc StreamAudio(StreamAudioRequest) returns (stream AudioChunk);
  
  // Get waveform data for visualization
  rpc GetWaveform(WaveformRequest) returns (WaveformData);
}

// ============================================================================
// MESSAGE DEFINITIONS
// ============================================================================

message Empty {}

// Conversion Request
message ConversionRequest {
  string source_path = 1;
  string voice = 2;
  float speed = 3;
  ConversionOptions options = 4;
}

message ConversionOptions {
  string output_format = 1;      // "m4b", "mp3", "wav"
  string mp3_bitrate = 2;        // "128k", "192k", "256k"
  bool normalize_volume = 3;
  float target_dbfs = 4;
  int32 chunk_size = 5;
  bool enable_parallel = 6;
  int32 max_parallel_chapters = 7;
}

// Progress Streaming
message ConversionProgress {
  enum Stage {
    IDLE = 0;
    PARSING = 1;
    CHUNKING = 2;
    CLEANING = 3;
    SYNTHESIZING = 4;
    ENCODING = 5;
    PACKAGING = 6;
    COMPLETE = 7;
    ERROR = 8;
    CANCELLED = 9;
  }
  
  Stage stage = 1;
  string stage_name = 2;           // Human-readable stage name
  
  // Progress
  int32 chapter_index = 3;
  int32 total_chapters = 4;
  int32 chunk_index = 5;
  int32 total_chunks = 6;
  float overall_progress = 7;      // 0.0 - 1.0
  
  // Timing
  string eta_seconds = 8;          // Formatted as "mm:ss" or "--:--"
  float elapsed_seconds = 9;
  
  // Current operation
  string current_chapter_title = 10;
  string message = 11;
  
  // Logs (batched for efficiency)
  repeated LogEntry logs = 12;
  
  // Result (only populated when stage == COMPLETE)
  ConversionResult result = 13;
  
  // Error details (only populated when stage == ERROR)
  ErrorDetails error = 14;
}

message LogEntry {
  int64 timestamp_unix_ms = 1;
  string level = 2;                // "DEBUG", "INFO", "PROCESS", "WARNING", "ERROR"
  string message = 3;
  string source = 4;               // Module/component name
}

message ConversionResult {
  bool success = 1;
  string title = 2;
  string author = 3;
  string output_path = 4;
  string output_format = 5;
  int64 total_duration_ms = 6;
  float processing_time_seconds = 7;
  repeated ChapterResult chapters = 8;
}

message ChapterResult {
  int32 chapter_number = 1;
  string chapter_title = 2;
  string mp3_path = 3;
  int64 duration_ms = 4;
  bool success = 5;
  string error = 6;
}

message ErrorDetails {
  string error_type = 1;
  string message = 2;
  string stack_trace = 3;
  int32 chapter_index = 4;         // Which chapter failed (-1 if N/A)
}

// Cancel
message CancelRequest {
  string conversion_id = 1;        // Future: support multiple conversions
}

message CancelResponse {
  bool success = 1;
  string message = 2;
}

message ConversionStatus {
  bool is_running = 1;
  string current_stage = 2;
  float progress = 3;
}

// Voice Preview
message PreviewRequest {
  string voice = 1;
  float speed = 2;
  string text = 3;                 // Optional custom text
}

// Audio Streaming
message AudioChunk {
  bytes data = 1;
  int32 sample_rate = 2;
  string format = 3;               // "pcm_f32le", "mp3", etc.
  bool is_last = 4;
  float timestamp_seconds = 5;     // For sync
}

message StreamAudioRequest {
  string book_id = 1;
  int32 chapter_number = 2;
  int64 start_position_ms = 3;     // For seeking
}

// ============================================================================
// MODEL MANAGEMENT
// ============================================================================

message ModelStatus {
  TTSModelStatus tts_model = 1;
  CleanerModelStatus cleaner_model = 2;
}

message TTSModelStatus {
  bool is_loaded = 1;
  string model_name = 2;
  int64 vram_usage_bytes = 3;
  string device = 4;               // "cpu", "gpu", "metal"
}

message CleanerModelStatus {
  bool is_loaded = 1;
  string model_name = 2;
  int64 vram_usage_bytes = 3;
}

message LoadModelRequest {
  enum ModelType {
    TTS = 0;
    CLEANER = 1;
  }
  ModelType model_type = 1;
  string model_name = 2;           // Optional: specific model variant
}

message LoadProgress {
  float progress = 1;              // 0.0 - 1.0
  string stage = 2;                // "downloading", "loading", "warming_up"
  string message = 3;
  int64 bytes_downloaded = 4;
  int64 bytes_total = 5;
}

message UnloadModelRequest {
  enum ModelType {
    TTS = 0;
    CLEANER = 1;
    ALL = 2;
  }
  ModelType model_type = 1;
}

// ============================================================================
// LIBRARY MANAGEMENT
// ============================================================================

message ListBooksRequest {
  int32 page_size = 1;
  string page_token = 2;           // For pagination
  string sort_by = 3;              // "created_at", "title", "author"
  bool ascending = 4;
}

message ListBooksResponse {
  repeated Book books = 1;
  string next_page_token = 2;
  int32 total_count = 3;
}

message Book {
  string id = 1;
  string title = 2;
  string author = 3;
  string cover_image_path = 4;
  int64 created_at_unix = 5;
  int64 duration_ms = 6;
  int32 chapter_count = 7;
  string source_path = 8;
  string output_path = 9;
  repeated ChapterInfo chapters = 10;
}

message ChapterInfo {
  int32 number = 1;
  string title = 2;
  int64 duration_ms = 3;
  string audio_path = 4;
}

message GetBookRequest {
  string book_id = 1;
}

message DeleteBookRequest {
  string book_id = 1;
  bool delete_files = 2;           // Also delete audio files
}

// Waveform
message WaveformRequest {
  string book_id = 1;
  int32 chapter_number = 2;
  int32 sample_points = 3;         // Number of data points to return
}

message WaveformData {
  repeated float samples = 1;      // Normalized -1.0 to 1.0
  int64 duration_ms = 2;
  int32 sample_rate = 3;
}
